sdk/program/src/lib.rs

```diff
@@ -1,3 +1,650 @@
+//! ["cdylib"]: https://doc.rust-lang.org/reference/linkage.html
+//!
+//! Because multiple crates that are linked together cannot all define
+//! program entrypoints (see the [`entrypoint!`] documentation) a common
+//! convention is to use a [Cargo feature] called `no-entrypoint` to allow
+//! the program entrypoint to be disabled.
+//!
+//! [Cargo feature]: https://doc.rust-lang.org/cargo/reference/features.html
+//!
+//! The skeleton of a Solana program typically looks like:
+//!
+//! ```
+//! #[cfg(not(feature = "no-entrypoint"))]
+//! pub mod entrypoint {
+//!     use solana_program::{
+//!         account_info::AccountInfo,
+//!         entrypoint,
+//!         entrypoint::ProgramResult,
+//!         pubkey::Pubkey,
+//!     };
+//!
+//!     entrypoint!(process_instruction);
+//!
+//!     pub fn process_instruction(
+//!         program_id: &Pubkey,
+//!         accounts: &[AccountInfo],
+//!         instruction_data: &[u8],
+//!     ) -> ProgramResult {
+//!         // Decode and dispatch instructions here.
+//!         todo!()
+//!     }
+//! }
+//!
+//! // Additional code goes here.
+//! ```
+//!
+//! With a `Cargo.toml` file that contains
+//!
+//! ```toml
+//! [lib]
+//! crate-type = ["cdylib", "rlib"]
+//!
+//! [features]
+//! no-entrypoint = []
+//! ```
+//!
+//! Note that a Solana program must specify its crate-type as "cdylib", and
+//! "cdylib" crates will automatically be discovered and built by the `cargo
+//! build-bpf` command. Solana programs also often have crate-type "rlib" so
+//! they can be linked to other Rust crates.
+//!
+//! # On-chain vs. off-chain compilation targets
+//!
+//! Solana programs run on the [rbpf] VM, which implements a variant of the
+//! [eBPF] instruction set. Because this crate can be compiled for both on-chain
+//! and off-chain execution, the environments of which are significantly
+//! different, it extensively uses [conditional compilation][cc] to tailor its
+//! implementation to the environment. The `cfg` predicate used for identifying
+//! compilation for on-chain programs is `target_arch = "bpf"`, as in this
+//! example from the `solana-program` codebase that logs a message via a
+//! syscall when run on-chain, and via a library call when offchain:
+//!
+//! [rbpf]: https://github.com/solana-labs/rbpf
+//! [eBPF]: https://ebpf.io/
+//! [cc]: https://doc.rust-lang.org/reference/conditional-compilation.html
+//!
+//! ```
+//! pub fn sol_log(message: &str) {
+//!     #[cfg(target_arch = "bpf")]
+//!     unsafe {
+//!         sol_log_(message.as_ptr(), message.len() as u64);
+//!     }
+//!
+//!     #[cfg(not(target_arch = "bpf"))]
+//!     program_stubs::sol_log(message);
+//! }
+//! # mod program_stubs {
+//! #     pub(crate) fn sol_log(message: &str) { }
+//! # }
+//! ```
+//!
+//! This `cfg` pattern is suitable as well for user code that needs to work both
+//! on-chain and off-chain.
+//!
+//! `solana-program` and `solana-sdk` were previously a single crate. Because of
```

Comment:
Would be helpful to point out what does and does not work in more detail...or if we can, wes should clean up this discrepancies

Reply:
Can I do this one in a follow up pr?

Reply:
Sounds good